Border colors 2

Marc-Andre-Rivet opened the pull request:
This is a continuation of the previous border colors [PR](https://github.com/plotly/dash-table/pull/402). This takes the essential elements from the PR and while keeping the same underlying ideas that were developed there, makes the implementation a bit tighter.

- [x] use subset of `border**` CSS rules to style the table's borders (border, border-left, border-right, border-bottom, border-top) --> while only supporting a subset might feel limiting, it covers pretty much all cases with very little redundancy vs. the full rules set and simplifies implementation greatly
- [x] omit all `border**` CSS rules from general cell/data/headers/filter/ops styling
- [x] apply default styling
- [x] reconcile common borders between various types & with n_fixed_rows & n_fixed_columns (priority order: data > filter > headers > operations > fillers)

n_fixed_rows, n_fixed_columns cases are handled through CSS: https://github.com/plotly/dash-table/pull/424/files#diff-02b1c6154ad8311a2face953de21380dR159

headers, filters, data and operations are reconciled with priority
1. data
2. filters
3. headers
4. operations

Given that all the items above share style_cell and style_cell_conditional, and that in all cases these are the lowest weight styles applied,

p1 will win on p2 if:
- p1's edge is not the table's default and not from style_cell or style_cell_conditional
- p1's edge weight is greater than p2's edge weight -- the weight is equal to the source style's index

- [x] unit tests for edge matrices generation
- [x] visual tests for table rendering

Note: Percy diffs will be all over the place because of the pixel-differences between `box-shadow` and `border` rendering. We'll need to be careful with that one.

Marc-Andre-Rivet commented on src/core/type/index.ts:
```diff
@@ -1,2 +1,10 @@
+export type RequiredProp<T, R extends keyof T> = T[R];
+export type OptionalProp<T, R extends keyof T> = T[R] | undefined;
+
+export type PropOf<T, R extends keyof T> = R;
```
Just defining some types to do TypeScript magic later..

alexcjohnson commented on src/core/math/arrayZipMap.ts:
```diff
@@ -2,7 +2,16 @@ import * as R from 'ramda';
+    const mapArray = R.addIndex<T1, TR>(R.map);
+
+    return mapArray((iValue, i) => cb(iValue, i), a1);
+}
```
This seems like a lot of trouble to go through to replicate `a1.map(cb)` - what's the benefit?

Marc-Andre-Rivet replied:
None. Only benefit is keeping with a relatively consistent ramda usage.

Marc-Andre-Rivet commented on src/core/math/matrixZipMap.ts:
```diff
@@ -2,70 +2,109 @@ import * as R from 'ramda';
+            iRow
+        ), m1
+    );
+}
```
Adding a 1-matrix map -- otherwise all below is essentially dealing with undefined cases

Marc-Andre-Rivet commented on src/dash-table/components/CellFactory.tsx:
```diff
@@ -100,15 +102,28 @@ export default class CellFactory {
+                children: [c],
+                style: R.mergeAll([
+                    s,
+                    dataEdges && dataEdges.getStyle(i, j)
```
Here and above in ops, obtaining the calculated borders for this specific cell

Marc-Andre-Rivet commented on src/dash-table/components/ControlledTable/index.tsx:
```diff
@@ -735,13 +717,30 @@ export default class ControlledTable extends PureComponent<ControlledTableProps>
+            ...(n_fixed_rows ? ['dash-freeze-top'] : []),
+            ...(n_fixed_columns ? ['dash-freeze-left'] : []),
+            ...(style_as_list_view ? ['dash-list-view'] : []),
+            ...(empty[0][1] ? ['dash-empty-01'] : []),
```
Adding a few marker classes for various cases
- empty content & fixed contents
- no data, no columns, no filter

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+
+import { ControlledTableProps, VisibleColumns, IViewportOffset, Data } from './Table/props';
+
+export default class EdgeFactory {
```
This is the central change of this PR -- evaluates styles in isolation & reconciles styles for boundary cases (e.g. shared data / filter borders, etc.)

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+        );
+
+        return target;
+    }
```
Evaluates boundary horizontal edges between the target edges matrix and the "next" edges matrix.

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+        );
+
+        return target;
+    }
```
Evaluates boundary vertical edges between the target edges matrix and the "next" edges matrix.

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+        R.forEach(j =>
+            (
+                hNext.getWeight(iNext, j) > cutoffWeight ||
+                (hNext.getWeight(iNext, j) > hTarget.getWeight(iTarget, j))
```
For each edge, if the "next" edge has a weight beyond the cutoff (not style_cell or style_cell_conditional) or if it's below-cutoff weight is higher than the target's, clear the target's edge

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+            (
+                vNext.getWeight(i, jNext) > cutoffWeight ||
+                (vNext.getWeight(i, jNext) > vTarget.getWeight(i, jTarget))
+            ) && vTarget.setEdge(i, jTarget, undefined, -Infinity, true),
```
For each edge, if the "next" edge has a weight beyond the cutoff (not style_cell or style_cell_conditional) or if it's below-cutoff weight is higher than the target's, clear the target's edge

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,228 @@
+        style_header_conditional: Headers,
+        data: Data,
+        offset: IViewportOffset
+    ) => {
```
The overall edges calculation is expensive and cached.
The specific edges calculation are cached.
The style calculation for each section are cache.

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,216 @@
+
+        headerOpEdges = this.vReconcile(headerOpEdges, headerEdges, cutoffWeight);
+        filterOpEdges = this.vReconcile(filterOpEdges, filterEdges, cutoffWeight);
+        dataOpEdges = this.vReconcile(dataOpEdges, dataEdges, cutoffWeight);
```
As they are cached, these need to be treated as immutable even though they are not. The reconciliation step will clone if required, afterwards the edges are returned as a read-only interface

Marc-Andre-Rivet commented on src/dash-table/components/Table/Table.less:
```diff
@@ -1,183 +1,5 @@
 @import (reference) '~dash-table/style/reset.less';
 
-.inset-shadow(@color, @left, @top, @right, @bottom) {
```
The box shadow declaration/rules are not required anymore 🔪

Marc-Andre-Rivet commented on src/dash-table/components/Table/Table.less:
```diff
@@ -334,6 +156,28 @@
+                    }
+                }
+            }
+        }
```
Handles the boundary cases for n_fixed_rows and n_fixed_columns -- only apply if the neighbouring table fragment is not empty

Marc-Andre-Rivet commented on src/dash-table/dash/DataTable.js:
```diff
@@ -77,8 +77,9 @@ export const defaultProps = {
-    style_data_conditional: [],
     style_cell_conditional: [],
+    style_data_conditional: [],
+    style_filter_conditional: [],
```
default was not defined for `style_filter_conditional`

Marc-Andre-Rivet commented on src/dash-table/derived/edges/type.ts:
```diff
@@ -0,0 +1,197 @@
+        p => p.indexOf('border') === 0,
+        Array.from(py2jsCssProperties.values())
+    )
+);
```
For use later, will omit all `border**` rules from the styles applied to the table cells

Marc-Andre-Rivet commented on src/dash-table/derived/edges/type.ts:
```diff
@@ -0,0 +1,197 @@
+        rowsOrMatrix: number | EdgesMatrix,
+        columns?: number,
+        defaultEdge?: Edge
+    ) {
```
Fwiw, this is how a ctor is overloaded in typescript -- define n ctors with their own definitions and then declare one that encompasses them all. As TS generates absolutely no new code, the developer is responsible for making it all work.

alexcjohnson commented on tests/cypress/tests/unit/edges_test.ts:
```diff
@@ -0,0 +1,319 @@
+            { columns: 0, rows: 0 }
+        );
+
+        expect(res !== undefined).to.equal(false);
```
Nice triple negative 😏 
`expect(res).to.be.undefined`?

alexcjohnson commented on tests/cypress/tests/unit/edges_test.ts:
```diff
@@ -0,0 +1,319 @@
+        }
+    });
+
+    it('applies `borderLeft` overridden by higher precedence `border`', () => {
```
can we test the opposite case too, `borderLeft` overriding only the left with `border` still setting the other three sides

Marc-Andre-Rivet replied:
adding

alexcjohnson commented:
@Marc-Andre-Rivet the percy diffs look fine to me. There's one where the baseline and new image both look wrong: **DashTable/Width percentages: with frozen rows** - there's misalignment between headers and data. Your call if you want to investigate this now or just make an issue for it, but I would like to see screenshots of variable-width borders (a big difference like 1 vs 4 px) to verify they don't cause misalignment:
- thick border on the headers, thin borders on data
- thin borders everywhere except thick borders on one highlighted data cell
- the same two cases between fixed and movable columns.

Also, do we already have a test case for conditional styling by both row number and data value while also sorting/filtering? Can we include borders there?

alexcjohnson commented:
Oh I didn't look at the python percy diffs at first - two of these seem to have broken:
- **editor-with-configurable-options** - the row selection column got much smaller? I actually kind of like it smaller, not sure why it had the size it did, but it's a substantial change.
- **dataframe-updating-graph-fe** actually the headers were already given sizes unrelated to the filters and data, but the two auto-generated initial columns collapsed making it worse

Marc-Andre-Rivet commented:
I'm seeing issues with at least:
- filter styling with invalid column query
- active cell border

Need to add more tests for borders. I'm barely covering the basics what the visual tests atm

Marc-Andre-Rivet commented:
@alexcjohnson Update:

Main problematic visual change I still see in the storybook is:
https://percy.io/plotly/dash-table/builds/1815314/view/118167441/1280?mode=diff&browser=chrome&snapshot=118167441 -- with the last bottom border probably hidden by the dropdown, to be checked.

Not sure about the percy python diffs:
https://percy.io/plotly/dash-table-python-v0/builds/1815331/view/118168907/1280?mode=diff&browser=chrome&snapshot=118168907 -- only seems to be affecting the "operation" column -- I don't see a significant change in the storybook snapshots for the operation columns.

https://percy.io/plotly/dash-table-python-v0/builds/1815331/view/118168743/1280?mode=diff&browser=chrome&snapshot=118168743 -- alignment of the header rows is incorrect, assuming n_fixed_rows and/or n_fixed_columns is set.

Still haven't addressed the issues you found earlier, still need to add more tests for more scenarios.

Not satisfied with the current filter / invalid styling solution (CSS / !important) -- will probably be problematic when styled with higher weight data rows. Will see when I add tests for those..

Marc-Andre-Rivet commented:
Changed test slightly for https://percy.io/plotly/dash-table/builds/1815314/view/118167441/1280?mode=diff&browser=chrome&snapshot=118167441 -- made table a bit higher to it's clear that the border is rendering.. the change was due to the small row height change with the border vs. the inset-box shadow.

https://percy.io/plotly/dash-table-python-v0/builds/1815331/view/118168907/1280?mode=diff&browser=chrome&snapshot=118168907 and https://percy.io/plotly/dash-table-python-v0/builds/1815331/view/118168743/1280?mode=diff&browser=chrome&snapshot=118168743 -- these are Python/Percy test artefacts, somehow with the implementation changes the old and the new artefacts are different. Rendering seems fine when run locally.

Marc-Andre-Rivet commented:
Last known issue: Filter case, using CSS the left border gets overridden by border-collapse giving precedence to the previous cell's borderRight
![image](https://user-images.githubusercontent.com/8092993/57255152-12041580-7021-11e9-8a53-93367758c30d.png)

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+            ) && hTarget.setEdge(i, jTarget, hPrevious.getEdge(i, jPrevious), Infinity, true)
+            , R.range(0, hPrevious.rows)
+        );
+    }
```
Evaluates boundary edges between the previous and the target -- and overrides the target's value with the previous value if it does not have precedence. This handles the fixed columns/rows cases

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+            style_header,
+            style_cell_conditional,
+            style_header_conditional
+        );
```
Calculate all the applicable styles for the various sub-sections of the table

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+            headerRows,
+            headerStyles,
+            style_as_list_view
+        );
```
Calculate the edges for each section

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+        filterEdges = EdgeFactory.clone(filterEdges);
+        filterOpEdges = EdgeFactory.clone(filterOpEdges);
+        dataEdges = EdgeFactory.clone(dataEdges);
+        dataOpEdges = EdgeFactory.clone(dataOpEdges);
```
Clone everything before operating on it... we want to keep the previous memoized results clean

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+
+        this.vReconcile(headerOpEdges, headerEdges, cutoffWeight);
+        this.vReconcile(filterOpEdges, filterEdges, cutoffWeight);
+        this.vReconcile(dataOpEdges, dataEdges, cutoffWeight);
```
Reconcile the boundaries between the table sections

Marc-Andre-Rivet commented on src/dash-table/components/EdgeFactory.tsx:
```diff
@@ -0,0 +1,314 @@
+            this.vOverride(headerOpEdges, headerEdges, cutoffWeight);
+            this.vOverride(filterOpEdges, filterEdges, cutoffWeight);
+            this.vOverride(dataOpEdges, dataEdges, cutoffWeight);
+        }
```
Override styles when necessary around the fixed rows & columns

Marc-Andre-Rivet commented:
@alexcjohnson This is finally ready for another round. It's a bit of a mess by now though :(

Marc-Andre-Rivet commented:
Found three more issues while demoing this to @cldougl 

- ops columns don't apply style for conditionals `if > row_index`
- `if > row_index` does not take paging into consideration
- filter columns with invalid values get erased when a new valid value is entered
